File: tests/XMP/global-view/reduction/F/mult2.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=1000
!$xmp nodes p(4,*)
!$xmp template t1(N,N)
!$xmp distribute t1(block,block) onto p
  integer a(N,N), sa, result
!$xmp align a(i,j) with t1(i,j)

  sa = 1
!$xmp loop (i,j) on t1(i,j)
  do j=1, N
     do i=1, N
        if(i.eq.j .and. mod(i,100).eq.0) then
           a(i,j) = 2
        else
           a(i,j) = 1
        endif
     enddo
  enddo

!$xmp loop (i,j) on t1(i,j)
  do j=1, N
     do i=1, N
        sa = sa*a(i,j)
     enddo
  enddo
!$xmp reduction (*:sa)

  result = 0
  if(  sa .ne. 1024 ) then
     result = -1 ! NG
  endif

!$xmp reduction(+:result)
!$xmp task on p(1,1)
  if( result .eq. 0 ) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  endif
!$xmp end task

end program main

File: tests/XMP/global-view/reduction/C/mult2.c
#include <xmp.h>
#include <stdio.h>
#include <stdlib.h>   
static const int N=1000;
#pragma xmp nodes p[*][4]
#pragma xmp template t1[N][N]
#pragma xmp distribute t1[block][block] onto p
int a[N][N],sa=1;
int i,j,m,result=0;
#pragma xmp align a[i][j] with t1[i][j]

int main(void)
{
#pragma xmp loop on t1[i][j]
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      if((i==j)&&(i%100==0)){
	a[i][j] = 2;
      }else{
	a[i][j] = 1;
      }
    }
  }

#pragma xmp loop on t1[i][j]
  for(i=0;i<N;i++)
    for(j=0;j<N;j++)
      sa = sa*a[i][j];
#pragma xmp reduction (*:sa) 

  if(sa != 1024)
    result = -1;  // ERROR

#pragma xmp reduction(+:result)
#pragma xmp task on p[0][0]
  {
    if(result == 0){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }
 
  return 0;
}
